codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,348 @@
+                        public void run() {
+                            streamInfo.childChannel.streamTerminated = true;
+                            streamInfo.childChannel.unsafe().disconnect(
+                                streamInfo.childChannel.newPromise());
+                        }
+                    });
+                }
+            }
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayMessage goAway = new Http2GoAwayMessage(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            if (streamInfo != null) {
+                                streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                ctx.invoker().invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway);
+        }
+    }
+
+    class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        public InternalHttp2ConnectionHandler(Http2Connection connection,
+                Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+                Http2FrameListener listener) {
+            super(connection, frameReader, frameWriter, listener);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                Http2Exception.StreamException http2Ex) {
+            super.onStreamError(ctx, cause, http2Ex);
+            Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+            if (stream == null) {
+                return;
+            }
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            if (streamInfo == null) {
+                return;
+            }
+            streamInfo.childChannel.pipeline().fireExceptionCaught(http2Ex);
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            // Use a user event in order to circumvent read queue.
```

Comment:
> Why would the user expect to get a socket read event unless "they" do a read? It may make more sense for "connection" (or "parent" channel) type events (i.e. go_away) where I'm not sure what the typical interaction pattern will look like from a user point of view (the child channels are what the user will be most interested in).

That's a fair argument. But I know for gRPC we do want the RST_STREAM or GOAWAY immediately, and not having it would be a deal-breaker. We don't want to waste server resources for a stream that hasn't consumed its inbound buffers, as that is _exactly_ the stream where RST_STREAM is most useful.

I can easily disagree with calling it a "socket read event," because reset of stream is not a normal Netty event. When we describe what it means, semantically, we simply describe its arrival semantics. I.e., we don't claim it is a "socket read event."

For the "connection/parent" channel, the argument is a bit moot, since we only would fire the event in response do doing a read(). We can just choose whichever is more "natural."

I will note that when ALLOW_HALF_CLOSURE is enabled, the closure is signaled with the ChannelInputShutdownEvent user event, but it will only happen in response to a read() in the actual code.

> Also using 2 different mechanism to dispatch events and/or dispatching events out of order seems like a confusing default to have.

This argument, as stated, is a fallacy as right now channelRead(), channelRegistered(), handlerAdded(), channelActive(), channelInactive(), channelWritablilityChanged(), etc. are all independent events and don't share a queue (not even the event loop). channelRead() is special in that it has flow control and so has its own queue.